pkg/lightning/backend/local/job_worker.go

```diff
@@ -117,6 +136,236 @@ func (w *regionJobBaseWorker) run(ctx context.Context) error {
 	}
 }
 
+// doRunJob handles a regionJob and tries to convert it to ingested stage.
+// If non-retryable error occurs, it will return the error.
+// If retryable error occurs, it will return nil and caller should check the stage
+// of the regionJob to determine what to do with it.
+func (w *regionJobBaseWorker) doRunJob(ctx context.Context, job *regionJob) error {
+	for {
+		// the job might in wrote stage if it comes from retry
+		if job.stage == regionScanned {
+			// writes the data to TiKV and mark this job as wrote stage.
+			// we don't need to do cleanup for the pairs written to TiKV if encounters
+			// an error, TiKV will take the responsibility to do so.
+			// TODO: let client-go provide a high-level write interface.
+			res, err := w.writeFn(ctx, job)
+			if err != nil {
+				if !w.isRetryableImportTiKVError(err) {
+					return err
+				}
+				// currently only one case will restart write
+				if strings.Contains(err.Error(), "RequestTooNew") {
+					// TiKV hasn't synced the newest region info with PD, it's ok to
+					// rewrite without rescan.
+					job.convertStageTo(regionScanned)
+				} else {
+					job.convertStageTo(needRescan)
+				}
+				job.lastRetryableErr = err
+				log.FromContext(ctx).Warn("meet retryable error when writing to TiKV",
+					log.ShortError(err), zap.Stringer("job stage", job.stage))
+				return nil
+			}
+			if res.emptyJob {
+				job.convertStageTo(ingested)
```

Comment:
these `j.convertStageTo` are written outside of `writeFn`. But for `ingestFn` they are written inside `ingestFn`. It's not tidy.

Reply:
yes, it's hard to refactor this part in a single PR, i write a todo in `cloudRegionJobWorker.ingest`
```go
		// TODO, we should let outer logic handle stage transition, currently, OP
		//  worker need a lot of change before we can do it, will refactor it later.
```